### PR TITLE
Prow compatible jobs

### PR DESCRIPTION
**Pull request by kargakis**

### CODE REVIEW

File: sjb/actions/pull_request_sync.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
           <defaultValue></defaultValue>
         </hudson.model.StringParameterDefinition>""")
 
-_PR_SYNC_TITLE_TEMPLATE = Template("SYNC {{ repository | upper }} PULL REQUEST ${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}")
+_PR_SYNC_TITLE_TEMPLATE = Template("SYNC {{ repository | upper }} PULL REQUEST {% if len(env['PULL_REFS']) > 0 %}$PULL_REFS{% else %}${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}{% endif %}")
```
Comment (kargakis) [inline sjb/actions/pull_request_sync.py]: This doesn't seem to work because it is statically generated prior to pushing in Jenkins - @stevekuznetsov not sure how to achieve what I am trying to do here, ie. output the correct variable that is used during the stage execution.

File: sjb/actions/pull_request_sync.py [hunk 2]
```diff
@@ -12,8 +12,67 @@
           <defaultValue></defaultValue>
         </hudson.model.StringParameterDefinition>""")
 
-_PR_SYNC_TITLE_TEMPLATE = Template("SYNC {{ repository | upper }} PULL REQUEST ${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}")
-_PR_SYNC_ACTION_TEMPLATE = Template("""oct sync remote {{ repository }} --refspec "pull/${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}/head" --branch "pull-${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}" --merge-into "${{ '{' }}{{ repository | replace('-', '_') | upper }}_TARGET_BRANCH}" """)
+_PULL_NUMBER_PARAMETER_TEMPLATE = Template("""        <hudson.model.StringParameterDefinition>
+          <name>PULL_NUMBER</name>
+          <description>The pull-request in the openshift/origin repository to test. This is compatible with prow.</description>
```
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: These are supposed to be repo-agnostic

File: sjb/actions/pull_request_sync.py [hunk 3]
```diff
@@ -12,8 +12,67 @@
...
+#
+# master:97d901d,4:bcb00a1
+#
+# And for multiple pull requests that have been batched:
+#
+# master:97d901d,4:bcb00a1,6:ddk2tka
+refs = [r.split(':')[0] for r in sys.argv[1].split(',')][1:]
+pull_refs = []
+for ref in refs:
+    if len(ref.strip()) == 0:
```
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: Why is this necessary?

File: sjb/actions/pull_request_sync.py [hunk 4]
```diff
@@ -12,8 +12,67 @@
...
+        continue
+    pull_refs.append(ref)
+
+refstring = ''
+for i, pull_num in enumerate(pull_refs):
```
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]:
```py
print(",".join(pull_refs))
```

?

File: sjb/actions/pull_request_sync.py [hunk 5]
```diff
@@ -12,8 +12,67 @@
...
+chmod +x unravel-pull-refs.py
+
+if [[ -n "${PULL_REFS:-}" ]]; then
+  IFS=',' read -ra REFS <<< "$(./unravel-pull-refs.py $PULL_REFS)"
+  for ref in "${{ '{' }}REFS[@]{{ '}' }}"; do
+      oct sync remote {{ repository }} --refspec "pull/$ref/head" --branch "pull-$ref" --merge-into "${{ '{' }}{{ repository | replace('-', '_') | upper }}_TARGET_BRANCH}"
+   done
+elif [[ -n "${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID:-}" ]]; then
+  oct sync remote {{ repository }} --refspec "pull/${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}/head" --branch "pull-${{ '{' }}{{ repository | replace('-', '_') | upper }}_PULL_ID}" --merge-into "${{ '{' }}{{ repository | replace('-', '_') | upper }}_TARGET_BRANCH}"
+fi
```
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: Feel like we need a 

```sh
else
    echo "[ERROR] Either \$PULL_REFS or \${{ repository | replace('-', '_') | upper }}_TARGET_BRANCH must be set"
    exit 1
fi
```

File: sjb/actions/pull_request_sync.py [hunk 6]
```diff
@@ -12,8 +12,67 @@
...
+    if len(pull_refs) > i+1:
+        refstring = '{{ '{' }}{{ '}' }},'.format(refstring)
+print('{{ '{' }}{{ '}' }}'.format(refstring))
+SCRIPT
+chmod +x unravel-pull-refs.py
+
+if [[ -n "${PULL_REFS:-}" ]]; then
+  IFS=',' read -ra REFS <<< "$(./unravel-pull-refs.py $PULL_REFS)"
+  for ref in "${{ '{' }}REFS[@]{{ '}' }}"; do
+      oct sync remote {{ repository }} --refspec "pull/$ref/head" --branch "pull-$ref" --merge-into "${{ '{' }}{{ repository | replace('-', '_') | upper }}_TARGET_BRANCH}"
```
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: Why are we `--merge-into` the envar and not the first spec in the list?

File: sjb/actions/pull_request_sync.py [hunk 7]
```diff
@@ -12,8 +12,67 @@
...
+for i, pull_num in enumerate(pull_refs):
+    refstring = '{{ '{' }}{{ '}' }}'.format(refstring+pull_num)
+    if len(pull_refs) > i+1:
+        refstring = '{{ '{' }}{{ '}' }},'.format(refstring)
+print('{{ '{' }}{{ '}' }}'.format(refstring))
+SCRIPT
+chmod +x unravel-pull-refs.py
+
+if [[ -n "${PULL_REFS:-}" ]]; then
+  IFS=',' read -ra REFS <<< "$(./unravel-pull-refs.py $PULL_REFS)"
```
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: If you newline-delimit or space-delimit the output from that script you don't need `$IFS` and `read` here and can just:

```sh
for ref in ./unravel-pull-refs.py "${PULL_REFS}":
```

**Comment by kargakis** [comment]
This is working now:
[Job that uses PULL_ORIGIN_ID](https://ci.openshift.redhat.com/jenkins/view/All/job/test_pull_request_prow_check/12/parameters/)
[Job that uses PULL_REFS+PULL_NUMBER](https://ci.openshift.redhat.com/jenkins/view/All/job/test_pull_request_prow_check/10/parameters/)
[Job that was started by prow](https://ci.openshift.redhat.com/jenkins/view/All/job/test_pull_request_prow_check/13/parameters/)

**Comment by kargakis** [comment]
> Job that was started by prow

It seems that I am running a patched version that adds ORIGIN_PULL_ID :smiling_imp:

File: sjb/actions/pull_request_sync.py (hunk 3 above)
Comment (kargakis) [inline sjb/actions/pull_request_sync.py]: Handles trailing commas like `master:97d901d,4:bcb00a1,6:ddk2tka,`

File: sjb/actions/pull_request_sync.py (hunk 6 above)
Comment (kargakis) [inline sjb/actions/pull_request_sync.py]: We need to merge all PRs in the target branch (usually master) because all the tests run on that branch.

Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: Sorry - I meant to ask, if the first part of `$PULL_REFS` is `master:shasum` shouldn't we be getting the `--merge-into` branch target from that `master:` prefix and not from the `$<repo>_TARGET_BRANCH` envar?

File: sjb/actions/pull_request_sync.py (hunk 3 above)
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: Ew... why is that input not sanitary?!?!

Comment (kargakis) [inline sjb/actions/pull_request_sync.py]: I think the input we get from k8s is ok, I was just testing the script manually and hit this. I can make sure we validate this upstream.

File: sjb/actions/pull_request_sync.py (hunk 6 above)
Comment (kargakis) [inline sjb/actions/pull_request_sync.py]: Yeah, I guess we can reuse that ref...

File: sjb/actions/pull_request_sync.py (hunk 3 above)
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: Yeah, if we can ensure we have good input coming in I would love to see this python script simple:

```py
print( ",".join([r.split(':')[0] for r in sys.argv[1].split(',')][1:]) )
```

File: sjb/actions/pull_request_sync.py (hunk 6 above)
Comment (stevekuznetsov) [inline sjb/actions/pull_request_sync.py]: I'm not sure it would be correct if we did not

**Comment by kargakis** [comment]
@stevekuznetsov all comments addressed

New jobs with both envs set separately below:
https://ci.openshift.redhat.com/jenkins/view/All/job/test_pull_request_prow_check/27/parameters/
https://ci.openshift.redhat.com/jenkins/view/All/job/test_pull_request_prow_check/28/parameters/
